skimage/filters/_gaussian.py

```diff
@@ -113,8 +117,17 @@ def gaussian(image, sigma=1, output=None, mode='nearest', cval=0,
         if len(sigma) != image.ndim:
             sigma = np.concatenate((np.asarray(sigma), [0]))
     image = convert_to_float(image, preserve_range)
-    return ndi.gaussian_filter(image, sigma, mode=mode, cval=cval,
-                               truncate=truncate)
+    if output is None:
+        output = np.empty_like(image)
+    elif not isinstance(output, np.ndarray):
+        raise ValueError("Provided output argument is not a numpy array")
+    elif image.dtype != output.dtype:
+        msg = ("Invalid output datatype. A new array is assigned.")
+        output = np.empty_like(image)
+        warn(RuntimeWarning(msg))
```

Comment:
This is pretty strange and likely to go uncaught by a user. When somebody passes in the output array, they are doing so intentionally. The coding style:

```
output = func(...., output=output)
```
is the general case where this would not be a bug, but I think this is pretty terrible coding style.


IMO, users should either use:

```
output = func(...)
```


```
func(..., output=output)
```

Other libraries may wish to use the "bad syntax" to support users of either camp, but we shouldn't encourage end users to do so.

TLDR, if you don't like the data type provided by the user, raise an error.

IMO, just let scipy deal with that.

Reply:
Thank you for this comment. I absolutely agree with this point. This code style may lead to uncaught bugs, but I just I follow scipy logic.

Reply:
I am also agree with raising type error, when data type provided by user is not correspond with image's one. Unlike you other reviewer instead on warning. I followed the recommendation.